sdk/dotnet/Pulumi.Automation/PulumiFn.cs

```diff
@@ -104,6 +104,18 @@ public static PulumiFn Create<TStack>()
         /// <typeparam name="TStack">The <see cref="Pulumi.Stack"/> type.</typeparam>
         public static PulumiFn Create<TStack>(IServiceProvider serviceProvider)
             where TStack : Pulumi.Stack
+            => Create<TStack>(serviceProvider, typeof(TStack));
+
+        /// <summary>
+        /// Creates an inline (in process) pulumi program via a traditional <see cref="Pulumi.Stack"/> implementation.
+        /// <para/>
+        ///     When invoked, a new stack instance will be resolved based
+        ///     on the provided <typeparamref name="TStack"/> type parameter
+        ///     using the <paramref name="serviceProvider"/> and <paramref name="serviceType"/>.
+        /// </summary>
+        /// <typeparam name="TStack">The <see cref="Pulumi.Stack"/> type.</typeparam>
+        public static PulumiFn Create<TStack>(IServiceProvider serviceProvider, Type serviceType)
```

Comment:
Doesn't this still having a type argument give you the same issue when you want to dynamically pass in your stack type? Like you don't know the `TStack` at compile-time which is why you want `serviceType`?

I think in order to support this we need to drop the type argument, and re-name `serviceType` to `stackType` and document it as needing to be a type that is assignable to `Pulumi.Stack`. And then right after we verify that `serviceProvider is not null` we should ensure that `typeof(Pulumi.Stack).IsAssignableFrom(stackType)`. Then we would use reflection to call to `PulumiFn.Create<TStack>` using `stackType`.

Reply:
Or to avoid the reflection we could make a non-generic implementation of `PulumiFn<TStack>` that takes in a type instance.